### PR TITLE
crimson/osd/snaptrim_event: get SnapSetContext from head obc

### PR DESCRIPTION
**Pull request by xxhdx1985126**
clone obc doesn't have snapset context

Fixes: https://tracker.ceph.com/issues/59511

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by Matan-B — COMMENTED** [review]
LGTM, few comments:
Can you please change the commit/PR title to `crimson/osd/snaptrim_event:...`?
nit, I would also suggest splitting the changes here into 2 commits. First one is getting the ssc from the head obc while the second one is encoding and setting the head obc updated snapset where needed.

File: src/crimson/osd/osd_operations/snaptrim_event.cc [hunk 1]
```diff
@@ -252,10 +253,18 @@ SnapTrimObjSubEvent::remove_clone(
...
+  head_obc->ssc->snapset.clone_size.erase(last);
+  head_obc->ssc->snapset.clone_snaps.erase(last);
+
+  ceph::bufferlist bss;
+  encode(head_obc->ssc->snapset, bss);
+  txn.setattr(
+    pg->get_collection_ref()->get_cid(),
+    ghobject_t{head_obc->obs.oi.soid},
+    SS_ATTR,
+    bss);
```
Comment (Matan-B) [inline src/crimson/osd/osd_operations/snaptrim_event.cc:267]: IIUC, we will set the SS_ATTR in `update_head` later on.

Comment (Matan-B) [inline src/crimson/osd/osd_operations/snaptrim_event.cc:267]: IIUC, we will set the SS_ATTR in `update_head` later on.
If that is the case we can omit this section.

File: src/crimson/osd/osd_operations/snaptrim_event.cc [hunk 2]
```diff
@@ -322,12 +331,21 @@ void SnapTrimObjSubEvent::remove_head_whiteout(
...
+  ceph::bufferlist bss;
+  head_obc->ssc->snapset.clone_snaps[coid.snap] =
     std::vector<snapid_t>(new_snaps.rbegin(), new_snaps.rend());
+  encode(head_obc->ssc->snapset, bss);
+  txn.setattr(
+    pg->get_collection_ref()->get_cid(),
+    ghobject_t{head_obc->obs.oi.soid},
+    SS_ATTR,
+    bss);
+
```
Comment (Matan-B) [inline src/crimson/osd/osd_operations/snaptrim_event.cc:348]: Same as previous comment.

File: src/crimson/osd/osd_operations/snaptrim_event.cc (hunk 1 above)
Comment (xxhdx1985126) [reply src/crimson/osd/osd_operations/snaptrim_event.cc]: > nit, I would also suggest splitting the changes here into 2 commits. First one is getting the ssc from the head obc while the second one is encoding and setting the head obc updated snapset attributes where needed.

You are right. And it seems that all "SS_ATTR" encoding added in this PR is not necessary. I've removed all of them, all the rests are "getting ssc from head obc". Please take a look, again, thanks:-)

**Review by Matan-B — APPROVED** [review]

**Review by athanatos — APPROVED** [review]

**Comment by athanatos** [comment]
https://pulpito.ceph.com/sjust-2023-04-25_14:29:45-crimson-rados-wip-sjust-crimson-testing-2023-04-24-distro-default-smithi/

**Comment by Matan-B** [comment]
https://pulpito.ceph.com/matan-2023-04-27_09:08:38-crimson-rados-wip-58928-59511-test-distro-crimson-smithi/

**Comment by Matan-B** [comment]
@xxhdx1985126, is this still DNM? I included this PR in the last round of testing.

Once make check is up again, this PR can be merged based on:

https://pulpito.ceph.com/matan-2023-04-30_16:46:54-crimson-rados-wip-matanb-crimson-only-testing-30.4-distro-crimson-smithi/
- 13/19 passed.
- Dead jobs: 
`7258784` - infra
`7258785` - infra
- Failed, known:
`7258780` - https://tracker.ceph.com/issues/59165
`7258781` - New test (tracked #50457)
`7258792` - rbd_lock_and_fence
`7258796` - rbd_python_api_tests_old_format

**Comment by athanatos** [comment]
jenkins test make check

**Comment by xxhdx1985126** [comment]
> @xxhdx1985126, is this still DNM? I included this PR in the last round of testing.

No, sorry that I forgot to remove the DNM label, this PR is good to go:-)

**Comment by Matan-B** [comment]
jenkins test api

**Review by rzarzynski — APPROVED** [review]

**Pull request merged**